### PR TITLE
Implemented URN reference rewriting for Bundles.

### PR DESCRIPTION
**Pull request by piotrszul**
Resolves: #1434

### CODE REVIEW

**Review by johngrimes — CHANGES_REQUESTED** [review]
Looks great @piotrszul!

Just a few minor things I noticed.

**Comment by johngrimes** [comment]
Would you mind also taking a look at CVE-2023-20862, so that we can get the Docker image test green?

**Comment by johngrimes** [comment]
I've tested this out with some Bundle-based analysis, all seems to work perfectly now! 🏅

File: encoders/src/main/java/au/csiro/pathling/support/r4/R4FhirConversionSupport.java [hunk 1]
```diff
@@ -42,14 +49,52 @@ public String fhirType(IBase base) {
...
+    if (base instanceof Reference) {
+      final Reference reference = (Reference) base;
+      final Resource resource = (Resource) reference.getResource();
+      if (isURNReference(reference) && resource != null && resource.hasIdElement()) {
+        reference.setReference(resource.getIdElement().getValue());
+      }
+    }
+  }
+
+  private static void resolveURNReferences(Resource resource) {
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/main/java/au/csiro/pathling/support/r4/R4FhirConversionSupport.java:82]: Fixed.

File: library-api/src/test/java/au/csiro/pathling/examples/EncodeBundles.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    final PathlingContext ptc = PathlingContext.create(spark);
+
+    // final Dataset<Row> bundlesDataset = spark.read().option("wholetext", true)
+    //     .text(bundlesData.toString());
+    // bundlesDataset.show();
+    //
+    // final Dataset<Row> conditionsFromBundleDS = ptc.encodeBundle(bundlesDataset, "Condition");
+    // conditionsFromBundleDS.show();
+    //
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply library-api/src/test/java/au/csiro/pathling/examples/EncodeBundles.java:73]: This was some testing code that I have checked in by mistake.
Deleted.

File: library-api/src/test/java/au/csiro/pathling/examples/EncodeBundles.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Base;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.Property;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+
+import static java.util.function.Predicate.not;
+
+public class EncodeBundles {
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply library-api/src/test/java/au/csiro/pathling/examples/EncodeBundles.java:22]: This was some testing code that I have checked in by mistake.
Deleted.

File: encoders/src/test/java/au/csiro/pathling/support/r4/R4FhirConversionSupportTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.hl7.fhir.r4.model.Patient;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.StringType;
+import org.junit.jupiter.api.Test;
+
+class R4FhirConversionSupportTest {
+
+  @Test
+  void testRewritesResolvedURNReference() {
+    IBaseResource referenceResource = new Patient().setId(new IdType("Patient", "1234"));
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply encoders/src/test/java/au/csiro/pathling/support/r4/R4FhirConversionSupportTest.java:34]: Done.

**Review by johngrimes — APPROVED** [review]
Thanks, looks good.

**Pull request merged**